File: output/pdfs/ocr_vision.m
#import <Foundation/Foundation.h>
#import <Vision/Vision.h>
#import <ImageIO/ImageIO.h>
#import <CoreGraphics/CoreGraphics.h>

static int pageNumberFromPath(NSString *path, int fallback) {
    NSString *name = [[path lastPathComponent] stringByDeletingPathExtension];
    NSInteger idx = [name length] - 1;
    while (idx >= 0) {
        unichar c = [name characterAtIndex:idx];
        if (c < '0' || c > '9') {
            break;
        }
        idx--;
    }
    if (idx == (NSInteger)[name length] - 1) {
        return fallback;
    }
    NSString *digits = [name substringFromIndex:(NSUInteger)(idx + 1)];
    return [digits intValue];
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        if (argc < 3) {
            fprintf(stderr, "usage: ocr_vision output.json image1.png [image2.png ...]\n");
            return 2;
        }

        NSString *outputPath = [NSString stringWithUTF8String:argv[1]];
        NSMutableArray *items = [NSMutableArray array];

        for (int argIndex = 2; argIndex < argc; argIndex++) {
            NSString *path = [NSString stringWithUTF8String:argv[argIndex]];
            NSURL *url = [NSURL fileURLWithPath:path];
            CGImageSourceRef source = CGImageSourceCreateWithURL((__bridge CFURLRef)url, NULL);
            if (source == NULL) {
                fprintf(stderr, "failed to load image source: %s\n", argv[argIndex]);
                continue;
            }

            CGImageRef image = CGImageSourceCreateImageAtIndex(source, 0, NULL);
            if (image == NULL) {
                fprintf(stderr, "failed to load image: %s\n", argv[argIndex]);
                CFRelease(source);
                continue;
            }

            VNRecognizeTextRequest *request = [[VNRecognizeTextRequest alloc] init];
            request.recognitionLevel = VNRequestTextRecognitionLevelAccurate;
            request.usesLanguageCorrection = NO;
            request.recognitionLanguages = @[@"zh-Hans", @"en-US"];

            VNImageRequestHandler *handler = [[VNImageRequestHandler alloc] initWithCGImage:image options:@{}];
            NSError *ocrError = nil;
            if (![handler performRequests:@[request] error:&ocrError]) {
                fprintf(stderr, "ocr failed for %s: %s\n", argv[argIndex], [[ocrError description] UTF8String]);
                CGImageRelease(image);
                CFRelease(source);
                continue;
            }

            int page = pageNumberFromPath(path, argIndex - 1);
            for (VNRecognizedTextObservation *observation in request.results) {
                VNRecognizedText *candidate = [[observation topCandidates:1] firstObject];
                if (candidate == nil || candidate.string == nil) {
                    continue;
                }
                CGRect box = observation.boundingBox;
                [items addObject:@{
                    @"page": @(page),
                    @"text": candidate.string,
                    @"confidence": @(candidate.confidence),
                    @"x": @(box.origin.x),
                    @"y": @(box.origin.y),
                    @"width": @(box.size.width),
                    @"height": @(box.size.height)
                }];
            }

            CGImageRelease(image);
            CFRelease(source);
        }

        NSError *jsonError = nil;
        NSData *jsonData = [NSJSONSerialization dataWithJSONObject:items options:NSJSONWritingPrettyPrinted error:&jsonError];
        if (jsonData == nil) {
            fprintf(stderr, "failed to encode json: %s\n", [[jsonError description] UTF8String]);
            return 1;
        }
        if (![jsonData writeToFile:outputPath atomically:YES]) {
            fprintf(stderr, "failed to write output: %s\n", [outputPath UTF8String]);
            return 1;
        }
    }
    return 0;
}

File: output/pdfs/ocr_vision.swift
import Foundation
import Vision
import ImageIO

struct OCRItem: Codable {
    let page: Int
    let text: String
    let confidence: Float
    let x: Double
    let y: Double
    let width: Double
    let height: Double
}

func loadCGImage(_ path: String) -> CGImage? {
    let url = URL(fileURLWithPath: path)
    guard let source = CGImageSourceCreateWithURL(url as CFURL, nil) else {
        return nil
    }
    return CGImageSourceCreateImageAtIndex(source, 0, nil)
}

func pageNumber(from path: String, fallback: Int) -> Int {
    let name = URL(fileURLWithPath: path).deletingPathExtension().lastPathComponent
    if let match = name.range(of: #"\d+$"#, options: .regularExpression) {
        return Int(name[match]) ?? fallback
    }
    return fallback
}

let args = Array(CommandLine.arguments.dropFirst())
guard args.count >= 2 else {
    FileHandle.standardError.write(Data("usage: ocr_vision.swift output.json image1.png [image2.png ...]\n".utf8))
    exit(2)
}

let outputPath = args[0]
let imagePaths = Array(args.dropFirst())
var allItems: [OCRItem] = []

for (index, path) in imagePaths.enumerated() {
    guard let image = loadCGImage(path) else {
        FileHandle.standardError.write(Data("failed to load image: \(path)\n".utf8))
        continue
    }
    let page = pageNumber(from: path, fallback: index + 1)
    let request = VNRecognizeTextRequest()
    request.recognitionLevel = .accurate
    request.usesLanguageCorrection = false
    request.recognitionLanguages = ["zh-Hans", "en-US"]
    if #available(macOS 13.0, *) {
        request.revision = VNRecognizeTextRequestRevision3
    }

    let handler = VNImageRequestHandler(cgImage: image, options: [:])
    do {
        try handler.perform([request])
    } catch {
        FileHandle.standardError.write(Data("ocr failed for \(path): \(error)\n".utf8))
        continue
    }

    let observations = request.results ?? []
    for observation in observations {
        guard let candidate = observation.topCandidates(1).first else {
            continue
        }
        let box = observation.boundingBox
        allItems.append(
            OCRItem(
                page: page,
                text: candidate.string,
                confidence: candidate.confidence,
                x: Double(box.origin.x),
                y: Double(box.origin.y),
                width: Double(box.width),
                height: Double(box.height)
            )
        )
    }
}

let encoder = JSONEncoder()
encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
let data = try encoder.encode(allItems)
try data.write(to: URL(fileURLWithPath: outputPath))
